keiko-sql/src/main/kotlin/com/netflix/spinnaker/q/sql/SqlQueue.kt

```diff
@@ -0,0 +1,743 @@
+      }
+    }
+
+    fire(QueuePolled)
+  }
+
+  override fun push(message: Message, delay: TemporalAmount) {
+    val fingerprint = message.hashV2()
+    val ulid = ULID().nextValue()
+    val deliveryTime = atTime(delay)
+
+    message.setAttribute(
+      message.getAttribute() ?: AttemptsAttribute()
+    )
+
+    withRetry(RetryCategory.WRITE) {
+      jooq.transaction { config ->
+        val txn = DSL.using(config)
+
+        txn.insertInto(messagesTable)
+          .set(idField, ulid.toString())
+          .set(fingerprintField, fingerprint)
+          .set(bodyField, mapper.writeValueAsString(message))
+          .onDuplicateKeyUpdate()
+          .set(bodyField, MySQLDSL.values(bodyField) as Any)
+          .execute()
+
+        txn.insertInto(queueTable)
+          .set(idField, ULID().nextMonotonicValue(ulid).toString())
+          .set(fingerprintField, fingerprint)
+          .set(deliveryField, deliveryTime)
+          .set(lockedField, "0")
+          .onDuplicateKeyUpdate()
+          .set(deliveryField, MySQLDSL.values(deliveryField) as Any)
+          .execute()
+      }
+    }
+
+    fire(MessagePushed(message))
+  }
+
+  override fun reschedule(message: Message, delay: TemporalAmount) {
+    val fingerprint = message.hashV2()
+
+    withRetry(RetryCategory.WRITE) {
+      val rows = jooq.update(queueTable)
+        .set(deliveryField, atTime(delay))
+        .where(fingerprintField.eq(fingerprint))
+        .execute()
+
+      if (rows == 1) {
+        log.debug("Rescheduled message: $message, fingerprint: $fingerprint to deliver in $delay")
+        fire(MessageRescheduled(message))
+      } else {
+        log.warn("Failed to reschedule message: $message, fingerprint: $fingerprint, not found " +
+          "on queue")
+        fire(MessageNotFound(message))
+      }
+    }
+  }
+
+  override fun ensure(message: Message, delay: TemporalAmount) {
+    val fingerprint = message.hashV2()
+    var missing = false
+
+    withRetry(RetryCategory.WRITE) {
+      jooq.transaction { config ->
+        val txn = DSL.using(config)
+
+        val queueRows = txn.select(fingerprintField)
+          .from(queueTable)
+          .where(fingerprintField.eq(fingerprint))
+          .limit(1)
+          .execute()
+
+        if (queueRows == 0) {
+          val unackedRows = txn.select(fingerprintField)
+            .from(unackedTable)
+            .where(fingerprintField.eq(fingerprint))
+            .limit(1)
+            .execute()
+
+          if (unackedRows == 0) {
+            missing = true
+          }
+        }
+      }
+    }
+
+    if (missing) {
+      log.debug(
+        "Pushing ensured message onto queue as it does not exist in queue or unacked tables"
+      )
+      push(message, delay)
+    }
+  }
+
+  private fun expireStaleLocks() {
+    val now = clock.instant().toEpochMilli()
+    val minMs = now.minus(TimeUnit.SECONDS.toMillis(lockTtlSeconds.toLong()))
+    val minUlid = ULID().nextValue(minMs).toString()
```

Comment:
🤔 This seems a bit exact? Shouldn't we look back further than `now - lockTtlSeconds`? What if a lock falls beyond the min here?

Reply:
`idField.lt(minUlid)` means we're initially looking at all currently locked rows originally inserted prior to `now - lockTtlSeconds`.